### DISCUSSION TITLE
v6.6.0以后的反向代理设置

### DISCUSSION THREAD

**Discussion by YuxuanZuo** [jeessy2/ddns-go#1129] [Q&A] [answered]
请问大家v6.6.0以后如何设置带路径的反向代理？比如把`http://127.0.0.1:9876`代理到`https://example.com/ddns`
以前用以下的代理配置是可以实现这个效果的，但是自从更新了webui以后会一直跳转到`https://example.com/login`，而正确的地址应该是`https://example.com/ddns/login`：
```
location ^~ /ddns/ {
    proxy_pass http://127.0.0.1:9876/; 
    proxy_set_header Host $host; 
    proxy_set_header X-Real-IP $remote_addr; 
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for; 
    proxy_set_header REMOTE-HOST $remote_addr; 
    proxy_set_header Upgrade $http_upgrade; 
    proxy_set_header Connection "upgrade"; 
    proxy_set_header X-Forwarded-Proto $scheme; 
    proxy_http_version 1.1; 
    add_header Cache-Control no-cache; 
    add_header Strict-Transport-Security "max-age=31536000"; 
}
```
恳请各位大佬指点！

**Comment by jeessy2** [comment, marked as answer]
根据你的描述，更新到webui v6.6.0之后，反向代理配置似乎遇到了一些问题，导致请求路径发生了不正确的跳转。你提供的配置基本上是正确的，但需要确保路径匹配和代理设置没有问题。下面是一个经过改进的配置示例：

```nginx
location /ddns/ {
    proxy_pass http://127.0.0.1:9876/;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header REMOTE-HOST $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_http_version 1.1;
    add_header Cache-Control no-cache;
    add_header Strict-Transport-Security "max-age=31536000";
    rewrite /ddns/(.*) /$1 break;
}
```

几点说明：

1. **`rewrite`指令**：`rewrite /ddns/(.*) /$1 break;` 这一行用于重写URL路径，以确保转发请求时，目标服务器能够正确处理路径。

2. **路径匹配**：`location /ddns/` 表示以 `/ddns/` 开头的所有请求，这种方式确保了 `/ddns/` 下的所有子路径都能够被正确代理。

3. **路径重写**：`rewrite`指令中的 `break` 表示完成重写后立即停止处理后续的重写指令，直接进入下一个阶段。

你可以尝试以上配置，如果还有问题，请检查以下几点：

- 确保 `proxy_pass` 的URL结尾带有 `/`，这样可以确保路径的正确拼接。
- 确保你的后端服务（`http://127.0.0.1:9876`）可以正确处理被代理的路径。
- 检查你的后端服务是否有URL重定向（例如 `/login` 重定向到根路径），这些可能会导致路径跳转问题。

通过这些调整，应该可以解决你在更新webui之后遇到的反向代理路径问题。

**Reply by YuxuanZuo** [reply]
> 根据你的描述，更新到webui v6.6.0之后，反向代理配置似乎遇到了一些问题，导致请求路径发生了不正确的跳转。你提供的配置基本上是正确的，但需要确保路径匹配和代理设置没有问题。下面是一个经过改进的配置示例：
> 
> ```nginx
> location /ddns/ {
>     proxy_pass http://127.0.0.1:9876/;
>     proxy_set_header Host $host;
>     proxy_set_header X-Real-IP $remote_addr;
>     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header REMOTE-HOST $remote_addr;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_http_version 1.1;
>     add_header Cache-Control no-cache;
>     add_header Strict-Transport-Security "max-age=31536000";
>     rewrite /ddns/(.*) /$1 break;
> }
> ```
> 
> 几点说明：
> 
> 1. **`rewrite`指令**：`rewrite /ddns/(.*) /$1 break;` 这一行用于重写URL路径，以确保转发请求时，目标服务器能够正确处理路径。
> 2. **路径匹配**：`location /ddns/` 表示以 `/ddns/` 开头的所有请求，这种方式确保了 `/ddns/` 下的所有子路径都能够被正确代理。
> 3. **路径重写**：`rewrite`指令中的 `break` 表示完成重写后立即停止处理后续的重写指令，直接进入下一个阶段。
> 
> 你可以尝试以上配置，如果还有问题，请检查以下几点：
> 
> * 确保 `proxy_pass` 的URL结尾带有 `/`，这样可以确保路径的正确拼接。
> * 确保你的后端服务（`http://127.0.0.1:9876`）可以正确处理被代理的路径。
> * 检查你的后端服务是否有URL重定向（例如 `/login` 重定向到根路径），这些可能会导致路径跳转问题。
> 
> 通过这些调整，应该可以解决你在更新webui之后遇到的反向代理路径问题。

感谢大佬的回复，rewrite规则看起来确实可行，但是请问可以考虑在后续版本中增加程序内部路由路径修改的支持吗？因为在群晖和一些特殊的系统下，nginx配置的修改可能并不是很方便

**Reply by jeessy2** [reply]
不支持，暂时没必要。

**Reply by rdndsy** [reply]
> 根据你的描述，更新到webui v6.6.0之后，反向代理配置似乎遇到了一些问题，导致请求路径发生了不正确的跳转。你提供的配置基本上是正确的，但需要确保路径匹配和代理设置没有问题。下面是一个经过改进的配置示例：
> 
> ```nginx
> location /ddns/ {
>     proxy_pass http://127.0.0.1:9876/;
>     proxy_set_header Host $host;
>     proxy_set_header X-Real-IP $remote_addr;
>     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header REMOTE-HOST $remote_addr;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_http_version 1.1;
>     add_header Cache-Control no-cache;
>     add_header Strict-Transport-Security "max-age=31536000";
>     rewrite /ddns/(.*) /$1 break;
> }
> ```
> 
> 几点说明：
> 
> 1. **`rewrite`指令**：`rewrite /ddns/(.*) /$1 break;` 这一行用于重写URL路径，以确保转发请求时，目标服务器能够正确处理路径。
> 2. **路径匹配**：`location /ddns/` 表示以 `/ddns/` 开头的所有请求，这种方式确保了 `/ddns/` 下的所有子路径都能够被正确代理。
> 3. **路径重写**：`rewrite`指令中的 `break` 表示完成重写后立即停止处理后续的重写指令，直接进入下一个阶段。
> 
> 你可以尝试以上配置，如果还有问题，请检查以下几点：
> 
> * 确保 `proxy_pass` 的URL结尾带有 `/`，这样可以确保路径的正确拼接。
> * 确保你的后端服务（`http://127.0.0.1:9876`）可以正确处理被代理的路径。
> * 检查你的后端服务是否有URL重定向（例如 `/login` 重定向到根路径），这些可能会导致路径跳转问题。
> 
> 通过这些调整，应该可以解决你在更新webui之后遇到的反向代理路径问题。

完全按上面配置后还是不能正常访问 版本 v6.5.0

**Reply by charley008** [reply]
https://github.com/jeessy2/ddns-go/issues/1137#issuecomment-2216363450

**Comment by jeessy2** [comment]
这个是可以的，还处理了非标端口问题

```
location /ddns/ {
    proxy_pass http://your_ip:9876/;  # 将请求转发到后端服务

    # 替换响应头中的 Location 字段，使其指向二级目录
    proxy_redirect / $scheme://$http_host/ddns/;

    # 处理可能的路径问题
    sub_filter_types text/html application/javascript text/css;
    sub_filter 'href="/' 'href="/ddns/';
    sub_filter 'src="/' 'src="/ddns/';
    sub_filter_once off;

    # 其他代理设置
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
}
```
